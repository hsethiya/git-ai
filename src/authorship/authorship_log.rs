use crate::authorship::transcript::Message;
use crate::authorship::working_log::AgentId;
use serde::{Deserialize, Serialize};
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Author {
    pub username: String,
    pub email: String,
}

/// Represents either a single line or a range of lines
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum LineRange {
    Single(u32),
    Range(u32, u32), // start, end (inclusive)
}

impl LineRange {
    pub fn contains(&self, line: u32) -> bool {
        match self {
            LineRange::Single(l) => *l == line,
            LineRange::Range(start, end) => line >= *start && line <= *end,
        }
    }

    #[allow(dead_code)]
    pub fn overlaps(&self, other: &LineRange) -> bool {
        match (self, other) {
            (LineRange::Single(l1), LineRange::Single(l2)) => l1 == l2,
            (LineRange::Single(l), LineRange::Range(start, end)) => *l >= *start && *l <= *end,
            (LineRange::Range(start, end), LineRange::Single(l)) => *l >= *start && *l <= *end,
            (LineRange::Range(start1, end1), LineRange::Range(start2, end2)) => {
                start1 <= end2 && start2 <= end1
            }
        }
    }

    /// Remove a line or range from this range, returning the remaining parts
    #[allow(dead_code)]
    pub fn remove(&self, to_remove: &LineRange) -> Vec<LineRange> {
        match (self, to_remove) {
            (LineRange::Single(l), LineRange::Single(r)) => {
                if l == r {
                    vec![]
                } else {
                    vec![self.clone()]
                }
            }
            (LineRange::Single(l), LineRange::Range(start, end)) => {
                if *l >= *start && *l <= *end {
                    vec![]
                } else {
                    vec![self.clone()]
                }
            }
            (LineRange::Range(start, end), LineRange::Single(r)) => {
                if *r < *start || *r > *end {
                    vec![self.clone()]
                } else if *r == *start && *r == *end {
                    vec![]
                } else if *r == *start {
                    vec![LineRange::Range(*start + 1, *end)]
                } else if *r == *end {
                    vec![LineRange::Range(*start, *end - 1)]
                } else {
                    vec![
                        LineRange::Range(*start, *r - 1),
                        LineRange::Range(*r + 1, *end),
                    ]
                }
            }
            (LineRange::Range(start1, end1), LineRange::Range(start2, end2)) => {
                if *start2 > *end1 || *end2 < *start1 {
                    // No overlap
                    vec![self.clone()]
                } else {
                    let mut result = Vec::new();
                    // Left part
                    if *start1 < *start2 {
                        result.push(LineRange::Range(*start1, *start2 - 1));
                    }
                    // Right part
                    if *end1 > *end2 {
                        result.push(LineRange::Range(*end2 + 1, *end1));
                    }
                    result
                }
            }
        }
    }

    /// Convert a sorted list of line numbers into compressed ranges
    pub fn compress_lines(lines: &[u32]) -> Vec<LineRange> {
        if lines.is_empty() {
            return vec![];
        }

        let mut ranges = Vec::new();
        let mut current_start = lines[0];
        let mut current_end = lines[0];

        for &line in &lines[1..] {
            if line == current_end + 1 {
                current_end = line;
            } else {
                // End current range and start new one
                if current_start == current_end {
                    ranges.push(LineRange::Single(current_start));
                } else {
                    ranges.push(LineRange::Range(current_start, current_end));
                }
                current_start = line;
                current_end = line;
            }
        }

        // Add the last range
        if current_start == current_end {
            ranges.push(LineRange::Single(current_start));
        } else {
            ranges.push(LineRange::Range(current_start, current_end));
        }

        ranges
    }

    #[allow(dead_code)]
    pub fn expand(&self) -> Vec<u32> {
        match self {
            LineRange::Single(l) => vec![*l],
            LineRange::Range(start, end) => (*start..=*end).collect(),
        }
    }

    /// Shift line numbers by a given offset
    /// - For insertions: offset is positive (shift lines down/forward)
    /// - For deletions: offset is negative (shift lines up/backward)
    /// - insertion_point: the line number where the change occurred
    #[allow(dead_code)]
    pub fn shift(&self, insertion_point: u32, offset: i32) -> Option<LineRange> {
        match self {
            LineRange::Single(l) => {
                if *l >= insertion_point {
                    let new_line = (*l as i32 + offset) as u32;
                    Some(LineRange::Single(new_line))
                } else {
                    Some(LineRange::Single(*l))
                }
            }
            LineRange::Range(start, end) => {
                let new_start = if *start >= insertion_point {
                    (*start as i32 + offset) as u32
                } else {
                    *start
                };
                let new_end = if *end >= insertion_point {
                    (*end as i32 + offset) as u32
                } else {
                    *end
                };

                // Ensure the range is still valid
                if new_start <= new_end {
                    if new_start == new_end {
                        Some(LineRange::Single(new_start))
                    } else {
                        Some(LineRange::Range(new_start, new_end))
                    }
                } else {
                    None
                }
            }
        }
    }
}

impl fmt::Display for LineRange {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            LineRange::Single(l) => write!(f, "{}", l),
            LineRange::Range(start, end) => write!(f, "[{}, {}]", start, end),
        }
    }
}

/// Prompt session details stored in the top-level prompts map keyed by short hash (agent_id + tool)
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PromptRecord {
    pub agent_id: AgentId,
    pub human_author: Option<String>,
    pub messages: Vec<Message>,
    #[serde(default)]
    pub total_additions: u32,
    #[serde(default)]
    pub total_deletions: u32,
    #[serde(default)]
    pub accepted_lines: u32,
    #[serde(default)]
    pub overriden_lines: u32,
}
