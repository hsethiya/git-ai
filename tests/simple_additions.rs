use git_ai::git::test_utils::TmpRepo;
use git_ai::log_fmt::authorship_log::LineRange;
use insta::assert_debug_snapshot;

#[test]
fn test_simple_additions_empty_repo() {
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo.write_file("test.txt", "Line1\n", true).unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    file.append("Line 2\nLine 3\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    let blame = tmp_repo.blame_for_file(&file, None).unwrap();
    assert_debug_snapshot!(blame);
}

#[test]
fn test_simple_additions_with_base_commit() {
    let (tmp_repo, mut lines, _) = TmpRepo::new_with_base_commit().unwrap();

    lines
        .append("NEW LINEs From Claude!\nHello\nWorld\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.commit_with_message("next commit").unwrap();

    let blame = tmp_repo.blame_for_file(&lines, None).unwrap();

    assert_debug_snapshot!(blame);
}

#[test]
fn test_simple_additions_on_top_of_ai_contributions() {
    let (tmp_repo, mut lines, _) = TmpRepo::new_with_base_commit().unwrap();

    lines
        .append("NEW LINEs From Claude!\nHello\nWorld\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.commit_with_message("next commit 1").unwrap();

    lines.replace_range(34, 36, "HUMAN ON AI\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("next commit 2").unwrap();

    let blame = tmp_repo.blame_for_file(&lines, Some((30, 34))).unwrap();
    assert_debug_snapshot!(blame);
}

#[test]
fn test_simple_additions_new_file_not_git_added() {
    let tmp_repo = TmpRepo::new().unwrap();

    // Create a new file that hasn't been git added yet
    let mut file = tmp_repo
        .write_file(
            "new_file.txt",
            "Line 1 from test_user\nLine 2 from test_user\nLine 3 from test_user\n",
            false,
        )
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    file.append("Line 4 from AI\nLine 5 from AI\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Git add and commit
    let authorship_log = tmp_repo.commit_with_message("Initial commit").unwrap();

    // The file should only have the AI lines attributed (lines 4-5)
    // Because the file wasn't tracked when test_user's checkpoint ran
    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_ai_human_interleaved_line_attribution() {
    let (tmp_repo, mut file, _) = TmpRepo::new_with_base_commit().unwrap();

    // AI adds line
    file.append("AI Line 1\n").unwrap();
    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Human adds line
    file.append("Human Line 1\n").unwrap();
    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    // AI adds another line
    file.append("AI Line 2\n").unwrap();
    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.commit_with_message("Interleaved commit").unwrap();

    let blame = tmp_repo.blame_for_file(&file, None).unwrap();
    assert_debug_snapshot!(blame);
}

#[test]
fn test_simple_ai_then_human_deletion() {
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.txt", "Line 1\nLine 2\nLine 3\nLine 4\nLine 5\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    file.append("AI Line\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.commit_with_message("AI adds line").unwrap();

    // Human deletes all AI lines by replacing them with empty
    file.replace_range(6, 7, "").unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    let authorship_log = tmp_repo
        .commit_with_message("Human deletes AI line")
        .unwrap();

    // The authorship log should have no attestations since we only deleted lines
    assert_eq!(authorship_log.attestations.len(), 0);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_multiple_ai_checkpoints_with_human_deletions() {
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo.write_file("test.txt", "Base\n", true).unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // First AI session
    file.append("AI1 Line 1\nAI1 Line 2\n").unwrap();
    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Second AI session
    file.append("AI2 Line 1\nAI2 Line 2\n").unwrap();
    tmp_repo
        .trigger_checkpoint_with_ai("GPT-4", Some("gpt-4"), Some("openai"))
        .unwrap();

    // Human deletes first AI session's lines
    file.replace_range(2, 4, "").unwrap();
    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    let authorship_log = tmp_repo.commit_with_message("Complex commit").unwrap();

    // Should only have AI2's lines attributed
    assert_eq!(authorship_log.attestations.len(), 1);
    let file_att = &authorship_log.attestations[0];
    assert_eq!(file_att.entries.len(), 1); // Only one AI session

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_complex_mixed_additions_and_deletions() {
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file(
            "test.txt",
            "Line 1\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nLine 8\nLine 9\nLine 10\n",
            true,
        )
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI deletes lines 2-3 and replaces with new content
    file.replace_range(2, 4, "NEW LINE A\nNEW LINE B\nNEW LINE C\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // AI inserts at the end
    file.append("END LINE 1\nEND LINE 2\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    let authorship_log = tmp_repo.commit_with_message("Complex edits").unwrap();

    // Should have lines 2-4 and the last 2 lines attributed to AI
    assert_eq!(authorship_log.attestations.len(), 1);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_ai_adds_lines_with_unstaged_modifications() {
    // Test that unstaged lines are NOT included in the authorship log after commit
    let tmp_repo = TmpRepo::new().unwrap();

    // Start with a base file
    let mut file = tmp_repo.write_file("test.ts", "base_line\n", true).unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    // Create initial commit
    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI adds 5 lines and we stage them
    tmp_repo
        .append_and_stage_file(
            &mut file,
            "ai_line1\nai_line2\nai_line3\nai_line4\nai_line5\n",
        )
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Now add more AI lines that won't be staged
    tmp_repo
        .append_unstaged_file(&mut file, "unstaged_ai_line1\nunstaged_ai_line2\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Commit only the staged changes
    let authorship_log = tmp_repo
        .commit_staged_with_message("AI adds lines with unstaged modifications")
        .unwrap();

    // The authorship log should only include lines 2-6 (the staged AI lines)
    // Lines 7-8 (unstaged_ai_line1, unstaged_ai_line2) should NOT be in the authorship log
    assert_debug_snapshot!(authorship_log);

    // Verify the blame only includes the committed lines
    let blame = tmp_repo.blame_for_file(&file, Some((1, 6))).unwrap();
    assert_debug_snapshot!(blame);
}

#[test]
fn test_partial_staging_filters_unstaged_lines() {
    // Test where AI makes changes but only some are staged
    let tmp_repo = TmpRepo::new().unwrap();

    // Start with a base file
    let mut file = tmp_repo
        .write_file("partial.ts", "line1\nline2\nline3\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI modifies lines 2-3 and stage
    file.replace_range(2, 4, "ai_modified2\nai_modified3\n")
        .unwrap();
    tmp_repo.stage_file("partial.ts").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Now AI adds more lines that won't be staged
    tmp_repo
        .append_unstaged_file(&mut file, "unstaged_line1\nunstaged_line2\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Partial staging")
        .unwrap();

    // Should only include lines 2-3 (the modifications), not the unstaged additions
    assert_eq!(authorship_log.attestations.len(), 1);
    let entry = &authorship_log.attestations[0].entries[0];
    assert_eq!(entry.line_ranges, vec![LineRange::Range(2, 3)]);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_human_stages_some_ai_lines() {
    // Test where AI adds multiple lines but human only stages some of them
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\nline3\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI adds lines 4-8
    file.append("ai_line4\nai_line5\nai_line6\nai_line7\nai_line8\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Stage only lines 4-6 (first 3 AI lines)
    tmp_repo.stage_lines_from_file(&file, &[(4, 6)]).unwrap();

    // Human adds an unstaged line
    file.append("human_unstaged\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Partial AI commit")
        .unwrap();

    // Should only have lines 4-6 from AI
    assert_eq!(authorship_log.attestations.len(), 1);
    let entry = &authorship_log.attestations[0].entries[0];
    assert_eq!(entry.line_ranges, vec![LineRange::Range(4, 6)]);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_multiple_ai_sessions_with_partial_staging() {
    // Multiple AI sessions, but only one has staged changes
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // First AI session adds lines and they get staged
    file.append("ai1_line1\nai1_line2\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    tmp_repo.stage_file("test.ts").unwrap();

    // Second AI session adds lines but they DON'T get staged
    file.append("ai2_line1\nai2_line2\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("GPT", Some("gpt-4"), Some("windsurf"))
        .unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Commit first AI session only")
        .unwrap();

    // Should only have the first AI session's lines
    assert_eq!(authorship_log.attestations.len(), 1);
    assert_eq!(authorship_log.attestations[0].entries.len(), 1);

    // Verify the lines are 3-4
    let entry = &authorship_log.attestations[0].entries[0];
    assert_eq!(entry.line_ranges, vec![LineRange::Range(3, 4)]);

    // Verify the hash corresponds to the first AI session (Claude)
    let hash = &entry.hash;
    let prompt_record = authorship_log.metadata.prompts.get(hash).unwrap();
    assert_eq!(prompt_record.agent_id.tool, "cursor");

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_stage_specific_lines_only() {
    // Test staging specific non-contiguous lines
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\nline3\nline4\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI adds lines 5-10
    file.append("ai_line5\nai_line6\nai_line7\nai_line8\nai_line9\nai_line10\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Stage only lines 5-7 (contiguous range)
    tmp_repo.stage_lines_from_file(&file, &[(5, 7)]).unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Commit specific lines")
        .unwrap();

    // Should have lines 5-7
    assert_eq!(authorship_log.attestations.len(), 1);
    let entry = &authorship_log.attestations[0].entries[0];

    let lines: Vec<u32> = entry.line_ranges.iter().flat_map(|r| r.expand()).collect();
    assert_eq!(lines, vec![5, 6, 7]);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_stage_middle_lines_leave_edges_unstaged() {
    // AI adds lines, stage only the middle ones
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI adds 6 lines (lines 3-8)
    file.append("ai_line3\nai_line4\nai_line5\nai_line6\nai_line7\nai_line8\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Stage lines 3-5 (leave 6, 7, 8 unstaged)
    tmp_repo.stage_lines_from_file(&file, &[(3, 5)]).unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Commit middle lines")
        .unwrap();

    // Should have lines 3-5
    assert_eq!(authorship_log.attestations.len(), 1);
    let entry = &authorship_log.attestations[0].entries[0];
    assert_eq!(entry.line_ranges, vec![LineRange::Range(3, 5)]);

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_multiple_ai_sessions_with_line_level_staging() {
    // Multiple AI sessions with line-level staging
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // First AI session adds lines 3-5
    file.append("ai1_line3\nai1_line4\nai1_line5\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Second AI session adds lines 6-8
    file.append("ai2_line6\nai2_line7\nai2_line8\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("GPT", Some("gpt-4"), Some("windsurf"))
        .unwrap();

    // Stage lines 3-5 from first session (will capture all of AI1's changes)
    tmp_repo.stage_lines_from_file(&file, &[(3, 5)]).unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("First AI session only")
        .unwrap();

    // Should have 1 entry for the first AI session
    assert_eq!(authorship_log.attestations.len(), 1);
    assert_eq!(authorship_log.attestations[0].entries.len(), 1);

    // Verify it's attributed to Claude (first session)
    let hash = &authorship_log.attestations[0].entries[0].hash;
    let prompt_record = authorship_log.metadata.prompts.get(hash).unwrap();
    assert_eq!(prompt_record.agent_id.tool, "cursor");

    assert_debug_snapshot!(authorship_log);
}

#[test]
fn test_interleaved_staged_unstaged_hunks() {
    // Complex case with interleaved staged/unstaged content
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\nline3\nline4\nline5\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI modifies/adds throughout the file
    file.replace_range(2, 3, "ai_modified_line2\n").unwrap();
    file.replace_range(4, 5, "ai_modified_line4\n").unwrap();
    file.append("ai_line6\nai_line7\nai_line8\n").unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Stage lines 2 and 6-7 (leave line 4 and 8 unstaged)
    tmp_repo
        .stage_lines_from_file(&file, &[(2, 2), (6, 7)])
        .unwrap();

    let authorship_log = tmp_repo
        .commit_staged_with_message("Interleaved staging")
        .unwrap();

    // Should have lines 2, 6, 7
    assert_eq!(authorship_log.attestations.len(), 1);
    let file_att = &authorship_log.attestations[0];
    assert!(!file_att.entries.is_empty());

    // Verify line 5 is NOT in the ranges
    for entry in &file_att.entries {
        for range in &entry.line_ranges {
            match range {
                LineRange::Single(l) => {
                    assert_ne!(*l, 5, "Line 5 should not be attributed (it was unstaged)");
                }
                LineRange::Range(start, end) => {
                    if *start <= 5 && *end >= 5 {
                        panic!("Line 5 should not be in any range (it was unstaged)");
                    }
                }
            }
        }
    }

    assert_debug_snapshot!(authorship_log);

    let blame = tmp_repo.blame_for_file(&file, Some((1, 8))).unwrap();
    assert_debug_snapshot!(blame);
}

#[test]
fn test_unstaged_ai_lines_saved_to_working_log() {
    // Test that unstaged AI-authored lines are saved to the working log for the next commit
    let tmp_repo = TmpRepo::new().unwrap();

    let mut file = tmp_repo
        .write_file("test.ts", "line1\nline2\nline3\n", true)
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_author("test_user")
        .unwrap();

    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI adds lines 4-7
    file.append("ai_line4\nai_line5\nai_line6\nai_line7\n")
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("claude-3-sonnet"), Some("cursor"))
        .unwrap();

    // Stage only lines 4-5, leaving 6-7 unstaged
    tmp_repo.stage_lines_from_file(&file, &[(4, 5)]).unwrap();

    // Commit only the staged lines
    let authorship_log = tmp_repo
        .commit_staged_with_message("Partial AI commit")
        .unwrap();

    // The authorship log should only have lines 4-5
    assert_eq!(authorship_log.attestations.len(), 1);
    let file_att = &authorship_log.attestations[0];
    assert_eq!(file_att.entries.len(), 1);

    // Lines 6-7 should have been saved to the working log
    // Let's verify by staging them and committing
    tmp_repo.stage_lines_from_file(&file, &[(6, 7)]).unwrap();

    let second_authorship_log = tmp_repo
        .commit_staged_with_message("Commit remaining AI lines")
        .unwrap();

    // The second commit should also attribute lines 6-7 to the AI
    // because they were saved to the working log
    assert_eq!(second_authorship_log.attestations.len(), 1);
    let second_file_att = &second_authorship_log.attestations[0];
    assert!(!second_file_att.entries.is_empty());

    // Verify the same AI session hash is used (from working log)
    let first_hash = &file_att.entries[0].hash;
    let second_hash = &second_file_att.entries[0].hash;
    assert_eq!(
        first_hash, second_hash,
        "Both commits should be attributed to the same AI session"
    );

    assert_debug_snapshot!((authorship_log, second_authorship_log));
}

/// Test: New file with partial staging across two commits
/// AI creates a new file with many lines, stage only some, then commit the rest
#[test]
fn test_new_file_partial_staging_two_commits() {
    let tmp_repo = TmpRepo::new().unwrap();
    tmp_repo.commit_with_message("Initial commit").unwrap();

    // AI creates a brand new file with a long list
    let mut file = tmp_repo
        .write_file(
            "planets.txt",
            "Mercury\nVenus\nEarth\nMars\nJupiter\nSaturn\nUranus\nNeptune\nPluto (dwarf)\n",
            false, // Don't stage yet
        )
        .unwrap();

    tmp_repo
        .trigger_checkpoint_with_ai("Claude", Some("gpt-4"), Some("cursor"))
        .unwrap();

    // Stage only the first 5 lines (inner planets + Jupiter)
    tmp_repo.stage_lines_from_file(&file, &[(1, 5)]).unwrap();

    // First commit with partial staging
    let log1 = tmp_repo
        .commit_staged_with_message("Add inner planets and Jupiter")
        .unwrap();

    // Check first commit's authorship log
    assert_eq!(log1.attestations.len(), 1);
    assert_eq!(log1.attestations[0].file_path, "planets.txt");

    // Should have lines 1-5
    let entry1 = &log1.attestations[0].entries[0];
    assert_eq!(entry1.line_ranges, vec![LineRange::Range(1, 5)]);

    // Now stage the remaining lines
    tmp_repo.stage_file("planets.txt").unwrap();

    // Second commit
    let log2 = tmp_repo
        .commit_staged_with_message("Add outer planets")
        .unwrap();

    // Check second commit's authorship log
    assert_eq!(log2.attestations.len(), 1);

    // Should have lines 6-9 (the previously unstaged lines)
    let entry2 = &log2.attestations[0].entries[0];
    assert_eq!(entry2.line_ranges, vec![LineRange::Range(6, 9)]);

    // Verify both commits use the same AI session hash
    assert_eq!(entry1.hash, entry2.hash);

    // Verify with blame
    let blame1 = tmp_repo.blame_for_file(&file, Some((1, 5))).unwrap();
    let blame2 = tmp_repo.blame_for_file(&file, Some((6, 9))).unwrap();

    assert_debug_snapshot!((log1, log2, blame1, blame2));
}
