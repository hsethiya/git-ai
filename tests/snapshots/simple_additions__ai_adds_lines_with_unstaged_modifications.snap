---
source: tests/simple_additions.rs
expression: authorship_log
---
AuthorshipLogV3 {
    attestations: [
        FileAttestation {
            file_path: "test.ts",
            entries: [
                AttestationEntry {
                    hash: "1cee1d9",
                    line_ranges: [
                        Range(
                            2,
                            6,
                        ),
                    ],
                },
            ],
        },
    ],
    metadata: AuthorshipMetadata {
        schema_version: "authorship/3.0.0",
        base_commit_sha: "b82b8fa3b85de492123de23d6aa6268f47522bd8",
        prompts: {
            "1cee1d9": PromptRecord {
                agent_id: AgentId {
                    tool: "cursor",
                    id: "test_session_fixed",
                    model: "claude-3-sonnet",
                },
                human_author: Some(
                    "Test User <test@example.com>",
                ),
                messages: [],
                total_additions: 7,
                total_deletions: 0,
                accepted_lines: 7,
            },
        },
    },
}
